### PR TITLE
OADP-1236 Add Data Mover to Upstream e2e

### PR DESCRIPTION
**Pull request by kaovilai**
[OADP-1236](https://issues.redhat.com/browse/OADP-1236)

Closes https://github.com/migtools/volume-snapshot-mover/issues/129

- Add a test entry containing verification functions that 
  - Discover and Install latest volsync via Operator catalog source either from go-client or via cli per [official instruction](https://access.redhat.com/documentation/en-us/openshift_container_platform/4.10/html-single/operators/index#olm-installing-operator-from-operatorhub-using-cli_olm-installing-operators-in-namespace) and wait for it to be ready.
- check app namespace events 
  - check velero volumesnapshot/content events
  - check VSM volumesnapshot/content events
snapshot status
PVC events (data mover and volsync)
volsync pod status
replicationSource status
VSB status
VSM pod logs

Run backup and restore and check data per usual.

Additional fixes
- Don't delete openshift-adp namespace after failure.

### CODE REVIEW

File: Makefile [hunk 1]
```diff
@@ -330,7 +330,7 @@ deploy-olm:
 	IMG=$(THIS_OPERATOR_IMAGE) BUNDLE_IMG=$(THIS_BUNDLE_IMAGE) \
 		make docker-build docker-push bundle bundle-build bundle-push; \
 	rm -rf $(DEPLOY_TMP)
-	operator-sdk run bundle $(THIS_BUNDLE_IMAGE) --namespace $(OADP_TEST_NAMESPACE)
+	operator-sdk run bundle $(THIS_BUNDLE_IMAGE) --namespace $(OADP_TEST_NAMESPACE) --index-image=quay.io/operator-framework/opm:v1.23.0
```
Comment (eemcmullan) [inline Makefile:333]: +1000

Comment (eemcmullan) [inline Makefile:333]: +1000

File: tests/e2e/lib/dpa_helpers.go [hunk 2]
```diff
@@ -102,26 +108,32 @@ func (v *DpaCustomResource) Build(backupRestoreType BackupRestoreType) error {
...
 	case CSI:
 		dpaInstance.Spec.Configuration.Restic.Enable = pointer.Bool(false)
 		defaultPlugins[oadpv1alpha1.DefaultPluginCSI] = emptyStruct{}
-		featureFlags["EnableCSI"] = emptyStruct{}
+		veleroFeatureFlags["EnableCSI"] = emptyStruct{}
+	case CSIDataMover:
+		dpaInstance.Spec.Configuration.Restic.Enable = pointer.Bool(false)
+		defaultPlugins[oadpv1alpha1.DefaultPluginCSI] = emptyStruct{}
+		veleroFeatureFlags["EnableCSI"] = emptyStruct{}
+		dpaInstance.Spec.Features.DataMover.Enable = true
```
Comment (eemcmullan) [inline tests/e2e/lib/dpa_helpers.go:129]: Do we also need to set `dataMover.credentialName` ?

File: tests/e2e/backup_restore_suite_test.go [hunk 3]
```diff
@@ -139,14 +156,24 @@ var _ = Describe("AWS backup restore tests", func() {
 				log.Printf("Waiting for restic pods to be running")
 				Eventually(AreResticPodsRunning(namespace), timeoutMultiplier*time.Minute*3, time.Second*5).Should(BeTrue())
 			}
-			if brCase.BackupRestoreType == CSI {
+			if brCase.BackupRestoreType == CSI || brCase.BackupRestoreType == CSIDataMover {
```
Comment (eemcmullan) [inline tests/e2e/backup_restore_suite_test.go:159]: Not sure if this is relevant here, but data mover currently only supports aws and azure

**Comment by eemcmullan** [comment]
@kaovilai should this also include a check for data mover deployment once DPA with `dataMover.enable` is true?

File: tests/e2e/backup_restore_suite_test.go (hunk 3 above)
Comment (kaovilai) [reply tests/e2e/backup_restore_suite_test.go:159]: Thanks.

File: tests/e2e/lib/dpa_helpers.go (hunk 2 above)
Comment (kaovilai) [reply tests/e2e/lib/dpa_helpers.go:130]: I thought you can just use default credential name? but yes can set it.

Comment (kaovilai) [reply tests/e2e/lib/dpa_helpers.go:130]: resolved via 2704f28...a51f67d

File: tests/e2e/backup_restore_suite_test.go (hunk 3 above)
Comment (kaovilai) [reply tests/e2e/backup_restore_suite_test.go:159]: resolved via 2704f28...a51f67d

**Comment by kaovilai** [comment]
> @kaovilai should this also include a check for data mover deployment once DPA with dataMover.enable is true?

Addressed here.
https://github.com/openshift/oadp-operator/commit/a51f67d197f259c6c1c2632572f8d08fb4e59469#diff-a9f567f9b6083a8f8e52d9318806042225c729fa99bea2354e148eead33bdf9fR33

**Comment by kaovilai** [comment]
Fixed CreateOrUpdate to use `merge-patch` strategy

**Comment by eemcmullan** [comment]
> > @kaovilai should this also include a check for data mover deployment once DPA with dataMover.enable is true?
> 
> Addressed here. [a51f67d#diff-a9f567f9b6083a8f8e52d9318806042225c729fa99bea2354e148eead33bdf9fR33](https://github.com/openshift/oadp-operator/commit/a51f67d197f259c6c1c2632572f8d08fb4e59469#diff-a9f567f9b6083a8f8e52d9318806042225c729fa99bea2354e148eead33bdf9fR33)

Can we also check for the data mover deployment? This looks like it's checking for just volsync. The dm deployment name is `common.DataMover` and in the OADP namespace.

**Review by eemcmullan — APPROVED** [review]

**Comment by kaovilai** [comment]
I once again request a
/retest

**Comment by kaovilai** [comment]
data mover seems to work but timeout before finishing restore

**Comment by kaovilai** [comment]
still seeing timeouts after 12 minutes.

**Comment by kaovilai** [comment]
@eemcmullan @shubham-pampattiwar  These logs are still showing as of current quay.io/konveyor/volume-snapshot-mover:latest

```
    2022/08/31 05:51:35  info Waiting for volumesnapshotbackup to complete mysql-persistent/vsb-fcg6h. Retrying in 5s
    2022/08/31 05:51:40  info Waiting for volumesnapshotbackup to complete mysql-persistent/vsb-fcg6h. Retrying in 5s
    2022/08/31 05:51:40  info Waiting for volumesnapshotbackup to complete mysql-persistent/vsb-fcg6h. Retrying in 5s
    2022/08/31 05:51:40 error Timed out awaiting reconciliation of volumesnapshotbackup mysql-persistent/vsb-fcg6h
```

Which means [this commit](https://github.com/openshift/velero-plugin-for-csi/commit/55c90e88a7011cf440d35ac8d2a683bb6b170473#diff-0f38f067df1a3a5e5fb78bd16bfeb63f7c7c89524abc32b98a875b6152474bb4R320) isn't built?

**Comment by eemcmullan** [comment]
@kaovilai Are you able to get the status of VSB? I'm wondering if it's in progress or even started.

**Comment by shubham-pampattiwar** [comment]
@kaovilai The logs `Waiting for volumesnapshotbackup to complete` are not removed, the block is just transferred from plugin to velero controller, we still have these logs in place but they are in Velero controller where we wait for the VSBs to complete before marking the backup as complete. Also, +1 on @eemcmullan's suggestion. Please check the status of VSB.

**Comment by kaovilai** [comment]
Only two actually failed.. the rest of the failure are install flakes.

**Comment by kaovilai** [comment]
Also we shouldn't be wondering if vsb started. Do we not have a log about vsb started in velero pod?

**Comment by eemcmullan** [comment]
Logs concerning VSB will mostly be in the data mover pod

**Comment by kaovilai** [comment]
Whatever is doing "Waiting for volumesnapshotbackup to complete" is already doing get call on this status, might as well report the current status it got in the log.

**Comment by kaovilai** [comment]
Can't find the logging code in https://github.com/openshift/velero/search?q=%22Waiting+for+volumesnapshotbackup+to+complete%22

**Comment by shubham-pampattiwar** [comment]
@kaovilai https://github.com/openshift/velero/commit/2220a8b0903cb3d7e187dc4246910a71948ca25f#diff-e9082eacb8358a6fdea744e078da1d6a43391e2d9ebf05f5e956f3295dcdfdafR67

**Comment by kaovilai** [comment]
/retest
will have to attach to cluster to see what's going on.

**Comment by kaovilai** [comment]
sad news is I attached to 4.11 and cannot reproduce.

**Comment by kaovilai** [comment]
Definitely flake worthy content

**Comment by kaovilai** [comment]
Adding retries to CreateOrUpdate

**Comment by kaovilai** [comment]
[4.9 aws](https://storage.googleapis.com/origin-ci-test/pr-logs/pull/openshift_oadp-operator/782/pull-ci-openshift-oadp-operator-master-4.9-operator-e2e-aws/1565755765203931136/build-log.txt)
```
2022/09/02 19:49:05 Waiting for velero pod to be running
2022/09/02 19:49:05 pod: velero-6c54fdc649-qlsc4 is not yet running with status: {Pending [{Initialized False 0001-01-01 00:00:00 +0000 UTC 2022-09-02 19:48:59 +0000 UTC ContainersNotInitialized containers with incomplete status: [velero-plugin-for-aws]} {Ready False 0001-01-01 00:00:00 +0000 UTC 2022-09-02 19:48:59 +0000 UTC ContainersNotReady containers with unready status: [velero]} {ContainersReady False 0001-01-01 00:00:00 +0000 UTC 2022-09-02 19:48:59 +0000 UTC ContainersNotReady containers with unready status: [velero]} {PodScheduled True 0001-01-01 00:00:00 +0000 UTC 2022-09-02 19:48:59 +0000 UTC  }]    10.0.189.232 10.129.2.110 [{10.129.2.110}] 2022-09-02 19:48:59 +0000 UTC [{openshift-velero-plugin {nil nil &ContainerStateTerminated{ExitCode:0,Signal:0,Reason:Completed,Message:,StartedAt:2022-09-02 19:49:03 +0000 UTC,FinishedAt:2022-09-02 19:49:03 +0000 UTC,ContainerID:cri-o://de62f66394812850ec6ef40492e39ff5126b4acf1670ed6086d53c848e1a611f,}} {nil nil nil} true 0 quay.io/konveyor/openshift-velero-plugin:latest quay.io/konveyor/openshift-velero-plugin@sha256:0fc7a01249f0ff7e98267f10b8a9ab411950987b136013b6167e3778ecf6ec70 cri-o://de62f66394812850ec6ef40492e39ff5126b4acf1670ed6086d53c848e1a611f <nil>} {velero-plugin-for-aws {&ContainerStateWaiting{Reason:PodInitializing,Message:,} nil nil} {nil nil nil} false 0 quay.io/konveyor/velero-plugin-for-aws:konveyor-1.5   <nil>}] [{velero {&ContainerStateWaiting{Reason:PodInitializing,Message:,} nil nil} {nil nil nil} false 0 quay.io/konveyor/velero:konveyor-1.9   0xc000902d4f}] Burstable []}
I0902 19:49:11.501693   13148 request.go:665] Waited for 1.001481337s due to client-side throttling, not priority and fairness, request: GET:https://api.ci-op-y3h8m1kw-311e5.origin-ci-int-aws.dev.rhcloud.com:6443/apis/image.openshift.io/v1?timeout=32s
2022/09/02 19:49:12 Checking for bsl spec
2022/09/02 19:49:12 Waiting for restic daemonset to be deleted
2022/09/02 19:49:12 Checking if Restic daemonset has been deleted...
2022/09/02 19:49:13 Waiting for velero deployment to have expected plugins
2022/09/02 19:49:13 Checking for default plugins
2022/09/02 19:49:13 Waiting for velero deployment to have expected custom plugins
[38;5;10m• [SLOW TEST] [10.618 seconds][0m
[0mConfiguration testing for DPA Custom Resource [38;5;243mUpdating custom resource with new configuration [0mAWS With Region And S3ForcePathStyle should succeed[0m [38;5;204m[aws][0m
[38;5;243m/go/src/github.com/openshift/oadp-operator/tests/e2e/dpa_deployment_suite_test.go:466[0m
[38;5;243m------------------------------[0m
2022/09/02 19:49:15 Test case expected to error. Waiting for the error to show in DPA Status
2022/09/02 19:49:18 Printing events for namespace:  openshift-adp
2022/09/02 19:49:18 Running must gather for failed test - AWS Without Region And S3ForcePathStyle true should fail
I0902 19:49:22.255998   13148 request.go:665] Waited for 1.050045279s due to client-side throttling, not priority and fairness, request: GET:https://api.ci-op-y3h8m1kw-311e5.origin-ci-int-aws.dev.rhcloud.com:6443/apis/datamover.oadp.openshift.io/v1alpha1?timeout=32s
[38;5;214m•! [INTERRUPTED] [5.566 seconds][0m
Configuration testing for DPA Custom Resource
[38;5;243m/go/src/github.com/openshift/oadp-operator/tests/e2e/dpa_deployment_suite_test.go:20[0m
  Updating custom resource with new configuration
  [38;5;243m/go/src/github.com/openshift/oadp-operator/tests/e2e/dpa_deployment_suite_test.go:549[0m
    [38;5;214m[1m[It] AWS Without Region And S3ForcePathStyle true should fail[0m [38;5;204m[aws][0m
    [38;5;243m/go/src/github.com/openshift/oadp-operator/tests/e2e/dpa_deployment_suite_test.go:502[0m

  [38;5;214mInterrupted by Timeout

  Here's a stack trace of all running goroutines:
    goroutine 54 [running]:
    github.com/onsi/ginkgo/v2/internal/interrupt_handler.(*InterruptHandler).InterruptMessageWithStackTraces(0xc000787b80)
    	/go/pkg/mod/github.com/onsi/ginkgo/v2@v2.1.1/internal/interrupt_handler/interrupt_handler.go:203 +0x212
    github.com/onsi/ginkgo/v2/internal.(*Suite).runNode(_, {0x1f, 0x4, {0x1b48c8f, 0x38}, 0xc00062d0e0, {{0x22219dd, 0x51}, 0x1f6, {0x0, ...}, ...}, ...}, ...)
    	/go/pkg/mod/github.com/onsi/ginkgo/v2@v2.1.1/internal/suite.go:614 +0x95d
    github.com/onsi/ginkgo/v2/internal.(*group).attemptSpec(0xc0009b1218, 0x1, {{0xc00055e400, 0xc00050e930, 0x2}, 0x0})
    	/go/pkg/mod/github.com/onsi/ginkgo/v2@v2.1.1/internal/group.go:181 +0x1049
    github.com/onsi/ginkgo/v2/internal.(*group).run(0xc0009b1218, {0xc000a32880, 0x1, 0xc00058f798})
    	/go/pkg/mod/github.com/onsi/ginkgo/v2@v2.1.1/internal/group.go:303 +0x665
    github.com/onsi/ginkgo/v2/internal.(*Suite).runSpecs(0xc0002a8000, {0x1af1037, 0xe}, {0x2d04238, 0x0, 0x0}, {0xc0008b0c00, 0x34}, 0x0, {0xc000568400, ...})
    	/go/pkg/mod/github.com/onsi/ginkgo/v2@v2.1.1/internal/suite.go:300 +0x61b
    github.com/onsi/ginkgo/v2/internal.(*Suite).Run(_, {_, _}, {_, _, _}, {_, _}, _, {0x1dee530, ...}, ...)
    	/go/pkg/mod/github.com/onsi/ginkgo/v2@v2.1.1/internal/suite.go:83 +0x376
    github.com/onsi/ginkgo/v2.RunSpecs({0x1dbc440, 0xc000503a00}, {0x1af1037, 0xe}, {0x0, 0x0, 0x154fae70d027})
    	/go/pkg/mod/github.com/onsi/ginkgo/v2@v2.1.1/core_dsl.go:280 +0x9b3
    github.com/openshift/oadp-operator/tests/e2e_test.TestOADPE2E(0xc000503a00)
    	/go/src/github.com/openshift/oadp-operator/tests/e2e/e2e_suite_test.go:86 +0x16b
    testing.tRunner(0xc000503a00, 0x1bdcd30)
    	/usr/lib/golang/src/testing/testing.go:1259 +0x102
    created by testing.(*T).Run
    	/usr/lib/golang/src/testing/testing.go:1306 +0x35a
```

**Comment by kaovilai** [comment]
Waiting on https://github.com/openshift/velero/runs/8208617595

**Comment by kaovilai** [comment]
4.8 backup timeout of 20 minutes not long enough at [backup_restore_suite_test.go:233](https://prow.ci.openshift.org/view/gs/origin-ci-test/pr-logs/pull/openshift_oadp-operator/782/pull-ci-openshift-oadp-operator-master-4.8-operator-e2e-aws/1567158440692813824#1:build-log.txt%3A7400)

The velero pod have only started to wait for volumesnapshotcontent [<1 seconds ago.](https://prow.ci.openshift.org/view/gs/origin-ci-test/pr-logs/pull/openshift_oadp-operator/782/pull-ci-openshift-oadp-operator-master-4.8-operator-e2e-aws/1567158440692813824#1:build-log.txt%3A7224)

**Comment by kaovilai** [comment]
blocked by https://github.com/backube/volsync/issues/420 :/

**Comment by kaovilai** [comment]
```
        mongo-persistent:  error preparing volumesnapshots.snapshot.storage.k8s.io/mongo-persistent/velero-mongo-j5z96: rpc error: code = Aborted desc = plugin panicked: runtime error: index out of range [0] with length 0

```
https://prow.ci.openshift.org/view/gs/origin-ci-test/pr-logs/pull/openshift_oadp-operator/782/pull-ci-openshift-oadp-operator-master-4.11-operator-e2e-aws/1588756368683175936#1:build-log.txt%3A615

**Comment by kaovilai** [comment]
index out of range resolved with.
```
dataMoverImageFqin: '[quay.io/emcmulla/data-mover:fix1016](http://quay.io/emcmulla/data-mover:fix1016)' and csiPluginImageFqin: '[quay.io/spampatt/velero-plugin-for-csi:fix-1020-blocker-poll-11-8](http://quay.io/spampatt/velero-plugin-for-csi:fix-1020-blocker-poll-11-8)'
```

now issue is
```
"time=\"2022-11-09T03:59:04Z\" level=error msg=\"Namespace mysql-persistent, resource restore error: error preparing volumesnapshots.snapshot.storage.k8s.io/mysql-persistent/velero-mysql-nrwz2: rpc error: code = Aborted desc = plugin panicked: runtime error: invalid memory address or nil pointer dereference\" logSource=\"pkg/controller/restore_controller.go:531\" restore=openshift-adp/mysql-datamover-e2e-54f6332f-5fe2-11ed-bf31-0a580a8379fb",
```

File: Makefile (hunk 1 above)
Comment (kaovilai) [reply Makefile:374]: reverting this since we are already downloading this specific versions in a few lines prior already

**Comment by kaovilai** [comment]
duplicate of #955

**Pull request closed without merging**